secure_software_development_fundamentals.md

```diff
@@ -2411,17 +2431,121 @@ Of course, like any technique, if you use it wrongly then it won’t be secure.
+The PostgreSQL `libpq` C interface provides several functions, as
+explained in the [PostgreSQL (Command Execution Functions) documentation](https://www.postgresql.org/docs/current/libpq-exec.html):
 
-There are other approaches, of course. You can write your own escape code, but this is difficult to get correct, and typically a waste of time since there are usually existing libraries to do the job. You can also use stored procedures, which can also help prevent SQL injection, but using them correctly can be a little tricky so we emphasize prepared statements here.
+* `PQexec` directly runs a single string command and returns a result.
+* `PQexecParams` implements a parameterized statement.
+  Placeholders are represented in the command as `$1`, `$2`, etc.,
+  and the parameter values are supplied as separate parameters in the same call.
+* `PQprepare` implements a prepared statement.
+  It takes a statement with placeholders and
+  submits it to the database to be prepared.
+  Users can later use the separate `PQexecPrepared` call
+  to provide the placeholder parameter values and execute the resulting
+  command.
 
-Properly using prepared statement libraries makes it much easier to write secure code. In addition, they typically make code easier to read, automatically handle the variations between how databases escape things, and sometimes they are faster than doing metacharacter escapes yourself.
+The [OWASP Query Parameterization Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Query_Parameterization_Cheat_Sheet.html) and [Bobby Tables website](https://bobby-tables.com/) provide examples for a variety of ecosystems.
+
+
+##### Stored Procedures
+
+Many database systems support "stored procedures", that is,
+procedures embedded in the database itself.
+If you use stored procedures, and commands (such as queries)
+are dynamically constructed in them, then you can again have
+SQL injection vulnerabilities.
+
+Again, the best solution is usually to use a parameterized query mechanism
+when creating the dynamic command (e.g., the SQL) inside the
+stored procedure. Be careful; in some systems using them correctly
+can be a little tricky.
+
+For more information on using parameterized queries
+in stored procedures, see your library's documentation, the
+[OWASP Query Parameterization Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Query_Parameterization_Cheat_Sheet.html), and the
+[OWASP SQL Injection Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html)
+
+##### When Parameterized Interfaces Won't Work
```

Comment:
What are Parameterized Interfaces? We haven't defined them.

Reply:
Let's just use "parameterized statements" for consistency.